src/main/java/teammates/common/datatransfer/questions/FeedbackConstantSumQuestionDetails.java

```diff
@@ -99,6 +146,67 @@ public List<String> validateQuestionDetails() {
             errors.add(CONST_SUM_ERROR_DUPLICATE_OPTIONS);
         }
 
+        if (distributeToRecipients) {
+            return errors;
+        }
+
+        // the validations below are only for constSumOptions
+        int totalPoints = pointsPerOption ? points * constSumOptions.size() : points;
+        double evenPointDistribution = ((double) totalPoints) / ((double) constSumOptions.size());
+
+        if (hasMinPoint) {
+            if (minPoint == null) {
+                errors.add(CONST_MIN_POINT_ERROR_NULL);
+                return errors;
+            }
+
+            if (minPoint < 0) {
+                errors.add(CONST_MIN_POINT_ERROR_NEGATIVE);
+            }
+            if (minPoint > totalPoints) {
+                errors.add(CONST_MIN_POINT_ERROR_EXCEEDS_POINTS + totalPoints);
+            }
+
+            // calculates the upper bound for minPoint to satisfy the total distribution of points among options
+            int upperBound = (int) Math.floor(evenPointDistribution);
+            if (minPoint > upperBound) {
+                errors.add(CONST_MIN_POINT_ERROR_ABOVE_UPPER_BOUND + upperBound);
+            }
+        } else {
+            if (minPoint != null) {
+                errors.add(CONST_MIN_POINT_ERROR_NOT_RESET);
+            }
+        }
+
+        if (hasMaxPoint) {
+            if (maxPoint == null) {
+                errors.add(CONST_MAX_POINT_ERROR_NULL);
+                return errors;
+            }
+
+            if (maxPoint < 0) {
+                errors.add(CONST_MAX_POINT_ERROR_NEGATIVE);
+            }
+
+            if (maxPoint > totalPoints) {
+                errors.add(CONST_MAX_POINT_ERROR_EXCEEDS_POINTS + totalPoints);
+            }
+
+            // calculates the lower bound for maxPoint to satisfy the total distribution of points among options
+            int lowerBound = (int) Math.ceil(evenPointDistribution);
+            if (maxPoint < lowerBound) {
+                errors.add(CONST_MAX_POINT_ERROR_BELOW_LOWER_BOUND + lowerBound);
+            }
+        } else {
+            if (maxPoint != null) {
+                errors.add(CONST_MAX_POINT_ERROR_NOT_RESET);
+            }
+        }
```

Comment:
This part of logic seems repetitive, could we make it into a helper function for min/max check and pass in min or max as a param to that helper function? That also means instead of repeating similar messages one time for max and another time for min in `FeedbackConstantSumQuestionDetails.java` line 36-53, we could define the consts as String Templates using `${variable}` to reduce the repeating code.

Reply:
Sure! Will update!

Reply:
@ziqing26 Here are few options for us to refactor the code with. Please let me know which would be most suitable, or if there are is better way, please point it out. For the time being, I have temporarily updated the code with  option 2, because I believe that is less likely to give bugs.

Option 1:
``` java
    // Caller function
    public List<String> validateQuestionDetails() {

       ....
      
      commonBoundaryValidation(errors, minPoint, PointType.MIN);
      commonBoundaryValidation(errors, maxPoint, PointType.MAX);


        if (maxPoint != null && minPoint != null && minPoint > maxPoint) {
            errors.add(CONST_SUM_ERROR_MIN_GREATER_THAN_MAX);
        }

        return errors;
    }

    // Helper function
    private void commonBoundaryValidation(List<String> errors, Integer points, PointType type) {
        if (points == null) {
            return;
        }

        // the validations below are only for constSumOptions
        int totalPoints = pointsPerOption ? points * constSumOptions.size() : points;
        double evenPointDistribution = ((double) totalPoints) / ((double) constSumOptions.size());

        String input;
        if (type == PointType.MAX) {
            input = MAX_POINT_STRING;
        } else if (type == PointType.MIN) {
            input = MIN_POINT_STRING;
        } else {
            input = "";
        }

        if (points < 0) {
            errors.add(String.format(CONST_SUM_TEMPLATE_NEGATIVE, input));
        }

        if (points > totalPoints) {
            errors.add(String.format(CONST_SUM_TEMPLATE_EXCEEDS_POINTS, input, totalPoints));
        }

        // calculates the lower bound for maxPoint to satisfy the total distribution of points among options
        int lowerBound = (int) Math.ceil(evenPointDistribution);
        if (type == PointType.MAX && points < lowerBound) {
            errors.add(CONST_SUM_ERROR_MAX_POINT_BELOW_LOWER_BOUND + lowerBound);
            return;
        }

        // calculates the upper bound for minPoint to satisfy the total distribution of points among options
        int upperBound = (int) Math.floor(evenPointDistribution);
        if (type == PointType.MIN && points > upperBound) {
            errors.add(CONST_SUM_ERROR_MIN_POINT_ABOVE_UPPER_BOUND + upperBound);
        }
    }
```

Option 2:
``` java
    // Caller function
    public List<String> validateQuestionDetails() {

        ....

        // the validations below are only for constSumOptions
        int totalPoints = pointsPerOption ? points * constSumOptions.size() : points;
        double evenPointDistribution = ((double) totalPoints) / ((double) constSumOptions.size());

        if (minPoint != null) {
            commonBoundaryValidation(errors, minPoint, totalPoints, MIN_POINT_STRING);

            // calculates the upper bound for minPoint to satisfy the total distribution of points among options
            int upperBound = (int) Math.floor(evenPointDistribution);
            if (minPoint > upperBound) {
                errors.add(CONST_SUM_ERROR_MIN_POINT_ABOVE_UPPER_BOUND + upperBound);
            }
        }

        if (maxPoint != null) {
            commonBoundaryValidation(errors, maxPoint, totalPoints, MAX_POINT_STRING);

            // calculates the lower bound for maxPoint to satisfy the total distribution of points among options
            int lowerBound = (int) Math.ceil(evenPointDistribution);
            if (maxPoint < lowerBound) {
                errors.add(CONST_SUM_ERROR_MAX_POINT_BELOW_LOWER_BOUND + lowerBound);
            }
        }

        if (maxPoint != null && minPoint != null && minPoint > maxPoint) {
            errors.add(CONST_SUM_ERROR_MIN_GREATER_THAN_MAX);
        }

        return errors;
    }

     // Helper function
    private void commonBoundaryValidation(List<String> errors, int points, int totalPoints, String boundaryType) {
        if (points < 0) {
            errors.add(String.format(CONST_SUM_TEMPLATE_NEGATIVE, boundaryType));
        }

        if (points > totalPoints) {
            errors.add(String.format(CONST_SUM_TEMPLATE_EXCEEDS_POINTS, boundaryType, totalPoints));
        }
    }
```
Option 3 (without helper function but with code refactored by removing hasMaxPoint and hasMinPoint):
``` java
     // Caller function
    public List<String> validateQuestionDetails() {
       
        ....

        if (minPoint != null) {
            if (minPoint < 0) {
                errors.add(String.format(CONST_SUM_TEMPLATE_NEGATIVE,  MIN_POINT_STRING));
            }
            if (minPoint > totalPoints) {
                errors.add(String.format(CONST_SUM_TEMPLATE_EXCEEDS_POINTS, MIN_POINT_STRING, totalPoints));
            }

            // calculates the upper bound for minPoint to satisfy the total distribution of points among options
            int upperBound = (int) Math.floor(evenPointDistribution);
            if (minPoint > upperBound) {
                errors.add(CONST_MIN_POINT_ERROR_ABOVE_UPPER_BOUND + upperBound);
            }
        }

        if (maxPoint != null) {
            if (maxPoint < 0) {
                errors.add(String.format(CONST_SUM_TEMPLATE_NEGATIVE,  MAX_POINT_STRING))
            }

            if (maxPoint > totalPoints) {
               errors.add(String.format(CONST_SUM_TEMPLATE_EXCEEDS_POINTS, MAX_POINT_STRING, totalPoints));
            }

            // calculates the lower bound for maxPoint to satisfy the total distribution of points among options
            int lowerBound = (int) Math.ceil(evenPointDistribution);
            if (maxPoint < lowerBound) {
                errors.add(CONST_MAX_POINT_ERROR_BELOW_LOWER_BOUND + lowerBound);
            }
        }

        if (maxPoint != null && minPoint != null && minPoint > maxPoint) {
            errors.add(CONST_POINT_ERROR_MIN_GREATER_THAN_MAX);
        }

        return errors;
    }
```

Reply:
Thanks for the suggestions! I think I am fine with option 2 👍